src/hotspot/share/opto/vectornode.cpp

```diff
@@ -2083,6 +2084,114 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
+        return static_cast<const Type*>(TypeLong::make(num_elem));
+      default:
+        fatal("Unsupported vectortype (%s)", type2name(elem_bt));
+        return static_cast<const Type*>(nullptr);
+    }
+  };
+
+  auto make_integral_index_vec = [&](Node* index_vec) {
+    switch(elem_bt) {
+      case T_FLOAT:
+        return phase->transform(new VectorCastF2XNode(index_vec, TypeVect::make(T_INT, num_elem)));
+      break;
+      case T_DOUBLE:
+        return phase->transform(new VectorCastD2XNode(index_vec, TypeVect::make(T_LONG, num_elem)));
+      break;
+      default:
+        return index_vec;
+    }
+  };
+
+  auto get_integal_type = [&](BasicType elem_bt) {
+    switch(elem_bt) {
+      case T_FLOAT:  return T_INT;
+      case T_DOUBLE: return T_LONG;
+      default: return elem_bt;
+    }
+  };
+
+  BasicType integral_elem_bt = get_integal_type(elem_bt);
+  int opc = VectorSupport::vop2ideal(VectorSupport::VECTOR_OP_SUB, integral_elem_bt);
+  int sopc = VectorNode::opcode(opc, integral_elem_bt);
+
+  BoolTest::mask pred = BoolTest::lt;
+  ConINode* pred_node = (ConINode*)phase->makecon(TypeInt::make(pred));
+  Node* lane_cnt = phase->makecon(lane_count_type());
+  Node* bcast_lane_cnt_vec = phase->transform(VectorNode::scalar2vector(lane_cnt, num_elem, Type::get_const_basic_type(integral_elem_bt), false));
+  Node* integral_index_vec = make_integral_index_vec(index_vec);
+
+  // Comparison over integral vectors weeds out emitting additional
+  // instructions for checking special floating point values.
+  const TypeVect* vmask_type = TypeVect::makemask(integral_elem_bt, num_elem);
+  Node* mask = phase->transform(new VectorMaskCmpNode(pred, integral_index_vec, bcast_lane_cnt_vec, pred_node, vmask_type));
+
+  vmask_type = TypeVect::makemask(elem_bt, num_elem);
+  mask = phase->transform(new VectorMaskCastNode(mask, vmask_type));
+
+  Node* p1 = phase->transform(new VectorRearrangeNode(src1, integral_index_vec));
+  Node* normalized_index_vec = phase->transform(VectorNode::make(sopc, integral_index_vec, bcast_lane_cnt_vec, vect_type()));
+  Node* p2 = phase->transform(new VectorRearrangeNode(src2, normalized_index_vec));
+
+  return new VectorBlendNode(p2, p1, mask);
+}
+
+Node* VectorRearrangeNode::Ideal(PhaseGVN* phase, bool can_reshape) {
+  BasicType elem_bt = vect_type()->element_basic_type();
+  int num_elem = vect_type()->length();
+  if (in(2)->Opcode() != Op_VectorUnbox &&
+      in(2)->Opcode() != Op_VectorLoadShuffle &&
+      Matcher::match_rule_supported_vector(Op_VectorRearrange, num_elem, elem_bt) &&
+      Matcher::vector_indexes_needs_massaging(elem_bt, num_elem)) {
+    auto get_integal_type = [&](BasicType elem_bt) {
+      switch(elem_bt) {
+        case T_FLOAT:  return T_INT;
+        case T_DOUBLE: return T_LONG;
+        default: return elem_bt;
+      }
+    };
+    // Targets emulating unsupported permutation for certain vector types
+    // may need to message the indexes to match the users intent.
```

Comment:
```suggestion
    // may need to massage the indexes to match the users intent.
```

Reply:
This optimization for now seems quite specific to your `SelectFromTwoVectorNode::Ideal` lowering code. Can this conversion not be done there already?

What is the semantics of `VectorRearrangeNode`? Should its shuffle vector always be bytes, and we now violated that "for a quick second"? Or is it going to be generally the idea to create all sorts of shuffle types and then fix that up? But then why do we need the `vector_indexes_needs_massaging`?

Can you help me understand the concept/strategy behind this?

Reply:
Ok, IIRC variable index permutation instruction on every target expects shape conformance b/w data vector and permute index vector.  Rearrange expects indices to be passed throug shuffle,  idealization routines automatically injects a VectorLoadShuffle after loading indexes held in shuffle's backing storage i.e. a byte array. 

In all the cases apart from byte vector permute , VectorLoadShuffle expands the index byte lanes to match the data vector lane. So we always end up emitting a lane expansion instruction before permute instruction (scenario 1). Apart from usual expansions VectorLoadShuffle may also do additional magic for some targets where it may need to prune / massage the index vector if target does not support destination vector type (scenario 2).

For our case, new selectFrom accepts the indices though vectors which save redundant expansions, but to leverage existing backend support for scenario 2 we do target specific pruning